src/LagrangianParticleTracking/advect_particles.jl

```diff
@@ -0,0 +1,78 @@
+"""
+    enforce_boundary_conditions(x, xₗ, xᵣ, ::Bounded)
+
+If a particle with position `x` and domain `xₗ < x < xᵣ` goes through the edge of the domain
+along a `Bounded` dimension, put them back at the wall.
+"""
+@inline function enforce_boundary_conditions(::Bounded, x, xₗ, xᵣ, restitution)
+    x > xᵣ && return xᵣ - (x - xᵣ) * restitution
+    x < xₗ && return xₗ + (xₗ - x) * restitution
+    return x
+end
+
+"""
+    enforce_boundary_conditions(x, xₗ, xᵣ, ::Periodic)
+
+If a particle with position `x` and domain `xₗ < x < xᵣ` goes through the edge of the domain
+along a `Periodic` dimension, put them on the other side.
+"""
+@inline function enforce_boundary_conditions(::Periodic, x, xₗ, xᵣ, restitution)
+    x > xᵣ && return xₗ + (x - xᵣ)
+    x < xₗ && return xᵣ - (xₗ - x)
+    return x
+end
+
+@kernel function _advect_particles!(particles, restitution, grid::RegularCartesianGrid{FT, TX, TY, TZ}, Δt, velocities) where {FT, TX, TY, TZ}
+    p = @index(Global)
+
+    # Advect particles using forward Euler.
+    @inbounds particles.x[p] += interpolate(velocities.u, Face(), Cell(), Cell(), grid, particles.x[p], particles.y[p], particles.z[p]) * Δt
+    @inbounds particles.y[p] += interpolate(velocities.v, Cell(), Face(), Cell(), grid, particles.x[p], particles.y[p], particles.z[p]) * Δt
+    @inbounds particles.z[p] += interpolate(velocities.w, Cell(), Cell(), Face(), grid, particles.x[p], particles.y[p], particles.z[p]) * Δt
```

Comment:
How does this work given that `Particles` is not `mutable` (am I confuzzled?)

Reply:
Ah it all works since particle locations/properties are stored in a `StructArray`. `particles.x` is an array but `particles[1]` reconstructs a `Particle` and returns it.

Not completely sure how it works but I think it converts an "array of structs" into "arrays for each property" so it becomes easy to pull out all the x locations to save particle locations to disk. It also becomes easy to add/remove particles.